include/nlohmann/detail/conversions/from_json.hpp

```diff
@@ -323,22 +345,71 @@ void from_json(const BasicJsonType& j, ArithmeticType& val)
     }
 }
 
-template<typename BasicJsonType, typename A1, typename A2>
-void from_json(const BasicJsonType& j, std::pair<A1, A2>& p)
+template<typename BasicJsonType, typename A1, typename A2,
+         enable_if_t<std::is_default_constructible<std::pair<A1, A2>>::value, int> = 0>
+void from_json(BasicJsonType && j, std::pair<A1, A2>& p)
+{
+    if (JSON_HEDLEY_UNLIKELY(!j.is_array()))
+    {
+        JSON_THROW(type_error::create(302, "type must be array, but is " +
+                                      std::string(j.type_name())));
+    }
+
+    p = {std::forward<BasicJsonType>(j).at(0).template get<A1>(),
+         std::forward<BasicJsonType>(j).at(1).template get<A2>()
+        };
+}
+
+template < typename BasicJsonType, class A1, class A2,
+           enable_if_t < !std::is_default_constructible<std::pair<A1, A2>>::value, int > = 0 >
```

Comment:
What would the signature of this `from_json` function be? The second argument could be either `std::pair<A1, A2>` or `identity_tag<std::pair<A1, A2>>`, and I'm not seeing how to automatically deduce `A1` and `A2` when the second argument is e.g. `PairRelatedType`.

Or do you mean to have a single overload taking an `std::pair` and one taking `identity_tag<std::pair>` and then both of those dispatching to `from_json_pair_impl`? In that case, if the SFINAE is removed for the `detail::tag` overloads, then there will be two possible overloads available for `adl_serializer` to call (in case of default constructable types), so then that won't compile anymore due to ambiguity.

I'm probably just overlooking something...

Reply:
I mean something like this: https://github.com/theodelrieu/mgs/blob/master/codecs/include/mgs/codecs/basic_codec.hpp#L109

Reply:
I've pushed an update which removes some code duplication and uses `priority_tag` to dispatch to the preferred function. Is this what you had in mind?

Note that `adl_serializer` now potentially has two `from_json` functions. This is e.g. the case for all default-constructable `std::array`, `std::pair`, and `std::tuple`. This doesn't cause ambiguity because `json::get()` uses SFINAE to select the "object returning" `from_json()` in that case ([discarded `get()`](https://github.com/AnthonyVH/json/blob/23f462b598b9da7b30e2924af39f807c45deaa63/include/nlohmann/json.hpp#L2927) vs [chosen `get()`](https://github.com/AnthonyVH/json/blob/23f462b598b9da7b30e2924af39f807c45deaa63/include/nlohmann/json.hpp#L2979)). This choice seems pretty deliberate, and looking at the code it's a logical choice due to possible bypassing at least 1 move construction (at least pre-C++17, although with optimizations there's probably no difference).

While doing some testing I did notice that the "chosen" `get()` gets called when deserializing a `std::vector<float>`. I found that quite surprising. I get how this `adl_serializer::from_json()` could now exist due to my code changes, but looking at the existing code it's not clear to me which `detail::from_json` would get called in that case. Just something I found odd which I figured was worth mentioning.

Reply:
>Is this what you had in mind?

Yes, you can remove this line for the `priority_tag<0>` overload though:

`enable_if_t < !std::is_default_constructible<std::array<T, N>>::value, int > = 0 >`

This is the point of `priority_tag<0>`, it forces priority inside the overload set so you can just keep the `is_default_constructible` check for `std::pair<A1, A2>`.

>Note that adl_serializer now potentially has two from_json functions.

IIRC (it's been a while since I've looked at the code), there is support for such overloads for `adl_serializer` specializations. I don't see why there would not be one for the base template indeed.

>While doing some testing I did notice that the "chosen" get() gets called when deserializing a std::vector<float>.

Seems like a change of behavior. To be honest, the template code is quite dusty, I've learned quite a lot since I wrote it, and it's quite hard to figure out all that's going on inside...

For instance, the different `get()` overloads would be a nice case for `priority_tag`. All the rules could be refined using [emulated concepts](https://theodelrieu.github.io/mgs-docs/concepts_refresher.html#concept-emulation).

Reply:
I've updated some more code and fixed a potential issue with my `std::pair` deserialization. Is this OK?